text/0000-default-serializers-and-adapters.md

```diff
@@ -0,0 +1,53 @@
+
+## Motivation
+
+The adapter and serializer packages provide reference implementations and base classes that are not required for applications that implement their own following the required interfaces for adapters and serializers as defined in their respective base classes.  Deprecating them allows us to simplify the lookup pattern and remove automatic injection and registration of potentially unused classes.
+
+In addition to removing use of initializer injection, this takes a significant step toward simplifying the mental model for how to determine what adapter/serializer is in use. Removing the defaults forces app developers to be more cognizant about the type of application level concerns vs model-specific concerns; they will now need to explicitly define and use specific adapters/serializers. After this RFC lands, apps will always use an adapter/serializer you explicitly put into your application and the rule will be "the adapter matching modelName falling back to application".
+
+## Detailed design
+The injection of `-default` and `-json-api` serializers will be removed in the next major version (4.0). Since this changes some core assumptions we will deprecate the reliance on the existence of the defaults. All deprecation warnings will only be shown in Dev mode.
+
+##### deprecate -default serializer fallback in Store.serializerFor
+A deprecation warning will be shown when [no model, application or adapter serializer](https://github.com/emberjs/data/blob/67affb0eca7048a1a0edc856af46d1305cd1fc1d/packages/store/addon/-private/system/store.ts#L2909) has specified and the default must be used.
+
+##### deprecate Adapter.serializer and Adapter.defaultSerializer fallbacks
+A deprecation warning will be shown when accessing the [adapter's defaultSerializer](https://github.com/emberjs/data/blob/67affb0eca7048a1a0edc856af46d1305cd1fc1d/packages/store/addon/-private/system/store.ts#L2896). This will be distinct from the warning about using the application level default.
+
+##### deprecate Store.defaultAdapter (-json-api) and the -json-api adapter fallback behavior
+A deprecation warning will be shown when the [defaultAdapter](https://github.com/emberjs/data/blob/b0cf3225662bfb806cd0c02b55b763e37a319b32/packages/store/addon/-private/system/store.ts#L309) is accessed.
+
+##### deprecate Model.toJSON
+A deprecation warning will be shown when toJSON is called since it uses a serializer to create a JSON representation of [the model](https://guides.emberjs.com/release/models/customizing-serializers/#toc_customizing-serializers). The message will suggests using a `debugData` method to accomplish the same goal.
+
+## How we teach this
+
+Today we have extensive documentation about creating custom serializers, but we will need to update the guides to specify the desired serializer in [app/serializers/application.js](https://guides.emberjs.com/release/models/customizing-serializers/#toc_customizing-serializers)
```

Comment:
My question applies to both serializers and adapters but am using the word adapters for brevity.
What practices should we encourage for application authors that do need various versions of adapters?
The guides recommend extending a base adapter and then implementing the hooks you need in each subclass. In this proposed update would we recommend writing independent functions within the application adapter?
```js
// app/adapters/application.js
import DS from 'ember-data';

export default DS.JSONAPIAdapter.extend({
  urlForQueryRecord({ slug }, modelName) {
    let baseUrl = this.buildURL();
    if (modelName === 'user') {
      return `${baseUrl}/other-place/${encodeURIComponent(slug)}`;
    }
    return `${baseUrl}/${encodeURIComponent(slug)}`;
  }
});
```

Reply:
The adapters/serializers would be have them same when in use. This RFC is only about deprecating the default injection of serializers. As it stands now, apps don't have to specify a adapter/serializer for their application because ember-data makes one available by default. This deprecates that default behavior, so in the future apps will have to explicitly specify an adapter/serializer per model and/or for their application.

Reply:
Also note that in practice the number of applications that choose not to provide an application adapter/serializer seems to be fairly low